src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+                if (!_promiseIsActive)
+                {
+                    _promiseIsActive = true;
+                    _valueOrEndPromise.Reset();
+                }
+                _valueOrEndPromise.SetResult(false);
+            }
+        }
+
+        // Contract: WaitForNextAsync will either return a false (no value left) or a true (found a value) with the promise set (to signal an unreturned value is stored).
+        public ValueTask<bool> WaitForNextAsync()
+        {
+            // if we have a pending promise already (ie. the async code was started by last TryGetNext), then we just return it 
+            // if we don't have a pending promise, then call MoveNext()
+            if (!this._promiseIsActive || State == StateMachineStates.NotStartedStateMachine)
+            {
+                MoveNext();
+                // MoveNext always returns with a promise of a future value, reaching end state, or an immediately available value
+
+                // In the case where a value is immediately available, we can rely on the last state of _valueOrEndPromise (TODO confirm this)
+            }
+
+            return new ValueTask<bool>(this, _valueOrEndPromise.Version);
+        }
+
+        /// <summary>
+        /// You should only call this method once the promise from WaitForNextAsync completed with true.
+        /// </summary>
+        public int TryGetNext(out bool success)
+        {
+            if (_promiseIsActive)
+            {
+                // throw if you call TryGetNext before WaitForNextAsync allowed you to
+                if (_valueOrEndPromise.GetStatus(_valueOrEndPromise.Version) == ValueTaskSourceStatus.Pending) throw new Exception();
+
+                // throw if state == -1 (you should call WaitForNextAsync first, ie. when in starting state)
+                if (State == StateMachineStates.NotStartedStateMachine) throw new Exception("You should call WaitForNextAsync first");
+
+                // if this is the first TryGetNext call after WaitForNext, then we'll return the value we already have
+                _promiseIsActive = false; // clear the promise, so that the next call to TryGetNext can move forward
+            }
+            else
+            {
+                // otherwise, call MoveNext to get a value or a promise of one
+                MoveNext();
+                // MoveNext always returns with a promise of a future value, reaching end state, or an immediately available value
+            }
+
+            if (_promiseIsActive || State == StateMachineStates.FinishedStateMachine)
+            {
+                success = false;
+                return default;
+            }
+
+            // the machine is stopped and state has value, so return it (success)
+            success = true;
+            return _current;
+        }
+
+        public void SetStateMachine(IAsyncStateMachine stateMachine) { }
+
+        // TODO when can we re-use this, and when should we create a new instance?
```

Comment:
At a minimum, I think we should encode into the state whether the instance has ever been used.  The 99.9% case we care about is just reusing the object for the first enumeration of the enumerable, as it's rare to actually enumerate the same enumerable again.  So something like this:
https://github.com/dotnet/corefx/blob/ecf20bcd7ce610c6cd242441c44af4389295757d/src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSourceTests.cs#L160-L163